iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp

```diff
@@ -0,0 +1,134 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_utils/cxx/generic_raii.hpp"
+#include "test.hpp"
+
+using namespace ::testing;
+using namespace iox::cxx;
+
+class GenericRAII_test : public Test
+{
+  public:
+    void SetUp() override
+    {
+    }
+
+    void TearDown() override
+    {
+    }
+};
+
+TEST_F(GenericRAII_test, InitFunctionIsCalledInCtorWhenSet)
+{
+    bool hasCalledInit = false;
+    GenericRAII sut([&] { hasCalledInit = true; }, std::function<void()>());
+    EXPECT_TRUE(hasCalledInit);
+}
+
+TEST_F(GenericRAII_test, InitFunctionIsCalledInCtorWhenSetWithCleanupFunction)
+{
+    bool hasCalledInit = false;
+    bool hasCalledCleanup = false;
+    GenericRAII sut([&] { hasCalledInit = true; }, [&] { hasCalledCleanup = true; });
+    EXPECT_TRUE(hasCalledInit);
+}
+
+TEST_F(GenericRAII_test, CleanupFunctionIsCalledInDtor)
+{
+    bool hasCalledInit = false;
+    bool hasCalledCleanup = false;
+
+    {
+        GenericRAII sut([&] { hasCalledInit = true; }, [&] { hasCalledCleanup = true; });
+    }
+
+    EXPECT_TRUE(hasCalledInit);
+    EXPECT_TRUE(hasCalledCleanup);
+}
+
+TEST_F(GenericRAII_test, CleanupFunctionIsCalledInDtorWhenUsingCleanupOnlyCTor)
+{
+    bool hasCalledCleanup = false;
+
+    {
+        GenericRAII sut([&] { hasCalledCleanup = true; });
+    }
+
+    EXPECT_TRUE(hasCalledCleanup);
+}
+
+TEST_F(GenericRAII_test, CleanupFunctionIsCalledInDtorWithEmptyInitFunction)
+{
+    bool hasCalledCleanup = false;
+
+    {
+        GenericRAII sut(std::function<void()>(), [&] { hasCalledCleanup = true; });
+    }
+
+    EXPECT_TRUE(hasCalledCleanup);
+}
+
+TEST_F(GenericRAII_test, MoveCTorDoesNotCallCleanupFunctionOfOrigin)
+{
+    bool hasCalledCleanup = false;
+
+    GenericRAII sut([&] { hasCalledCleanup = true; });
+    GenericRAII sut2(std::move(sut));
+
+    EXPECT_FALSE(hasCalledCleanup);
+}
+
+TEST_F(GenericRAII_test, MoveConstructedDoesCallCleanupFunctionWhenDestroyed)
+{
+    bool hasCalledCleanup = false;
+
+    {
+        GenericRAII sut([&] { hasCalledCleanup = true; });
+        GenericRAII sut2(std::move(sut));
+    }
+
+    EXPECT_TRUE(hasCalledCleanup);
+}
+
+TEST_F(GenericRAII_test, MoveAssignmentCallsCleanup)
+{
+    bool hasCalledCleanup = false;
+    bool hasCalledCleanup2 = false;
+
+    GenericRAII sut([&] { hasCalledCleanup = true; });
+    GenericRAII sut2([&] { hasCalledCleanup2 = true; });
+
+    sut = std::move(sut2);
+
+    EXPECT_TRUE(hasCalledCleanup);
+}
+
+TEST_F(GenericRAII_test, MoveAssignedCallsCleanupWhenOutOfScope)
+{
+    bool hasCalledCleanup = false;
+    bool hasCalledCleanup2 = false;
+
+    {
+        GenericRAII sut([&] { hasCalledCleanup = true; });
+        GenericRAII sut2([&] { hasCalledCleanup2 = true; });
+
+        sut = std::move(sut2);
+    }
+
+    EXPECT_TRUE(hasCalledCleanup);
+    EXPECT_TRUE(hasCalledCleanup2);
+}
```

Comment:
same here, how do you know that the sut2 dtor doesn't set the cleanup flag

Reply:
counter